File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/ee8cdc58ccb008e8e7b395dc02609459.py
import sys
from collections import defaultdict
from math import comb

MOD = 10**9 + 7

class Ring:
    def __init__(self, value=0):
        self.value = value % MOD

    def __mul__(self, other):
        return Ring(self.value * other.value % MOD)

    def __add__(self, other):
        return Ring((self.value + other.value) % MOD)

    def __sub__(self, other):
        return Ring((self.value - other.value) % MOD)

    def __truediv__(self, other):
        return self * other.inv()

    def inv(self):
        return self.pow(MOD - 2)

    def pow(self, p):
        r = Ring(1)
        x = Ring(self.value)
        while p:
            if p & 1:
                r *= x
            x *= x
            p //= 2
        return r

    def __repr__(self):
        return str(self.value)

def fast_scan():
    return list(map(int, sys.stdin.readline().split()))

def fast_print(value):
    print(value)

def binom(n, k):
    if k > n:
        return Ring(0)
    return Ring(fact[n]) / (Ring(fact[n - k]) * Ring(fact[k]))

n = int(sys.stdin.readline())
fact = [1] * (n + 100)

for i in range(1, n + 100):
    fact[i] = fact[i - 1] * i % MOD

v = fast_scan()
cnt = defaultdict(int)

for x in v:
    for j in range(2, int(x**0.5) + 1):
        while x % j == 0 and (x // j) % j == 0:
            x //= j
            x //= j
    cnt[x] += 1

c = list(cnt.values())
dp = [Ring(0)] * (n + 10)
dp[0] = Ring(1)

for x in c:
    ndp = [Ring(0)] * (n + 10)
    for i in range(len(dp)):
        if dp[i].value:
            for j in range(x):
                dist = x - j
                ndp[i + dist] += dp[i] * binom(x - 1, j) * binom(i + dist, dist)
    dp = ndp

conds = n - len(c)
final = Ring(0)
sign = Ring(1)

for i in range(conds + 1):
    final += sign * dp[n - i]
    sign = Ring(-1) * sign

for x in c:
    final *= Ring(fact[x])

fast_print(final)

File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/ee8cdc58ccb008e8e7b395dc02609459.go
package main

import (
	"fmt"
	"math"
)

const MOD = 1e9 + 7

type Ring struct {
	value int64
}

func floorMod(a, b int64) int64 {
	a %= b
	if a < 0 {
		a += b
	}
	return a
}

func NewRing(value int64) Ring {
	return Ring{floorMod(value, MOD)}
}

func (r Ring) Pow(p int64) Ring {
	res := NewRing(1)
	x := r
	for p > 0 {
		if p&1 == 1 {
			res *= x
		}
		x *= x
		p >>= 1
	}
	return res
}

func (r Ring) Inv() Ring {
	return r.Pow(MOD - 2)
}

func (r *Ring) Mul(b Ring) {
	r.value *= b.value
	r.value = floorMod(r.value, MOD)
}

func (r Ring) Mul(b Ring) Ring {
	res := r
	res.Mul(b)
	return res
}

func (r *Ring) Add(b Ring) {
	r.value += b.value
	r.value -= (r.value >= MOD) * MOD
}

func (r Ring) Add(b Ring) Ring {
	res := r
	res.Add(b)
	return res
}

func (r *Ring) Sub(b Ring) {
	r.value -= b.value
	r.value += (r.value < 0) * MOD
}

func (r Ring) Sub(b Ring) Ring {
	res := r
	res.Sub(b)
	return res
}

func (r *Ring) Div(b Ring) {
	*r *= b.Inv()
}

func (r Ring) Div(b Ring) Ring {
	res := r
	res.Div(b)
	return res
}

func (r Ring) Equals(b Ring) bool {
	return r.value == b.value
}

func (r Ring) NotEquals(b Ring) bool {
	return r.value != b.value
}

func (r Ring) String() string {
	return fmt.Sprintf("%d", r.value)
}

var fact []Ring

func binom(n, k int) Ring {
	if k > n {
		return NewRing(0)
	}
	return fact[n].Div(fact[n-k].Mul(fact[k]))
}

func main() {
	var n int
	fmt.Scan(&n)

	fact = make([]Ring, n+100)
	fact[0] = NewRing(1)
	for i := 1; i < n+100; i++ {
		fact[i] = fact[i-1].Mul(NewRing(int64(i)))
	}

	v := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&v[i])
	}

	for i := 0; i < n; i++ {
		for j := 2; j*j <= v[i]; j++ {
			for v[i]%j == 0 && (v[i]/j)%j == 0 {
				v[i] /= j
				v[i] /= j
			}
		}
	}

	cnt := make(map[int]int)
	for _, x := range v {
		cnt[x]++
	}

	c := make([]int, 0, len(cnt))
	for _, pp := range cnt {
		c = append(c, pp)
	}

	dp := make([]Ring, n+10)
	dp[0] = NewRing(1)

	for _, x := range c {
		ndp := make([]Ring, n+10)
		for i := 0; i < len(dp); i++ {
			if dp[i].value != 0 {
				for j := 0; j < x; j++ {
					dist := x - j
					ndp[i+dist] = ndp[i+dist].Add(dp[i].Mul(binom(x-1, j)).Mul(binom(i+dist, dist)))
				}
			}
		}
		dp = ndp
	}

	conds := n - len(c)

	final := NewRing(0)
	sign := NewRing(1)
	for i := 0; i <= conds; i++ {
		final = final.Add(sign.Mul(dp[n-i]))
		sign = NewRing(-1).Mul(sign)
	}
	for _, x := range c {
		final = final.Mul(fact[x])
	}
	fmt.Println(final)
}
